otherlibs/win32unix/symlink.c

```diff
@@ -28,32 +28,80 @@
 #include <caml/osdeps.h>
 #include "unixsupport.h"
 
+#ifndef SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE
+#define SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE (0x2)
+#endif
+
 typedef BOOLEAN (WINAPI *LPFN_CREATESYMBOLICLINK) (LPWSTR, LPWSTR, DWORD);
 
+static BOOL is_initialized = FALSE;
 static LPFN_CREATESYMBOLICLINK pCreateSymbolicLink = NULL;
-static int no_symlink = 0;
+static BOOL no_symlinks_available = FALSE;
+static DWORD additional_symlink_flags = 0;
+
+// Developer Mode allows the creation of symlinks without elevation - see
+// https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createsymboliclinkw
+static BOOL IsDeveloperModeEnabled()
+{
+  HKEY hKey;
+  LSTATUS openKeyError, queryValueError;
+  DWORD developerModeRegistryValue, dwordSize = sizeof(DWORD);
+
+  openKeyError = RegOpenKeyExW(
+    HKEY_LOCAL_MACHINE,
+    L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\AppModelUnlock",
+    0,
+    KEY_READ | KEY_WOW64_64KEY,
+    &hKey
+  );
+  if (openKeyError != ERROR_SUCCESS) {
+    return FALSE;
+  }
+
+  queryValueError = RegQueryValueExW(
+    hKey,
+    L"AllowDevelopmentWithoutDevLicense",
+    NULL,
+    NULL,
+    (LPBYTE)&developerModeRegistryValue,
+    &dwordSize
+  );
+  RegCloseKey(hKey);
+  if (queryValueError != ERROR_SUCCESS) {
+    return FALSE;
+  }
+  return developerModeRegistryValue != 0;
+}
 
 CAMLprim value unix_symlink(value to_dir, value osource, value odest)
 {
   CAMLparam3(to_dir, osource, odest);
-  DWORD flags = (Bool_val(to_dir) ? SYMBOLIC_LINK_FLAG_DIRECTORY : 0);
+  DWORD flags;
   BOOLEAN result;
   LPWSTR source;
   LPWSTR dest;
   caml_unix_check_path(osource, "symlink");
   caml_unix_check_path(odest, "symlink");
 
 again:
-  if (no_symlink) {
+  if (no_symlinks_available) {
     caml_invalid_argument("symlink not available");
   }
 
-  if (!pCreateSymbolicLink) {
+  if (!is_initialized) {
     pCreateSymbolicLink = (LPFN_CREATESYMBOLICLINK)GetProcAddress(GetModuleHandle(L"kernel32"), "CreateSymbolicLinkW");
-    no_symlink = !pCreateSymbolicLink;
+    no_symlinks_available = !pCreateSymbolicLink;
+
+    if (!no_symlinks_available && IsDeveloperModeEnabled()) {
+      additional_symlink_flags = SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE;
+    }
+
+    is_initialized = TRUE;
```

Comment:
Why do you need a separate flag `is_initialized`? Isn't it equivalent to having `pCreateSymbolicLink != NULL` like the existing check?

Reply:
I don't technically need it but it seemed more readable to me. It requires additional reading and understanding to know that `pCreateSymbolicLink` is set during initialization. `is_initialized` conveys that directly through its name and the reader can continue without having to jump anywhere else.

Also with the two-step init of setting both `pCreateSymbolicLink` _and_ `additional_symlink_flags`, we could, at some point when this code changes again, end up in a situation where `pCreateSymbolicLink` is set but `additional_symlink_flags` haven't been set for some reason. Then we would assume that it was initialized when in fact it wasn't and we'd have a hard to debug bug that breaks developer mode symlinks.

Don't you like it?

Reply:
I think it adds an extra cognitive burden on the reader. If there were two separate pieces of state that can evolve differently, then it would be OK, but in this case, I think it would be better to leave it as before.

Reply:
Removed `is_initialized` again